### PR TITLE
Make enchantment mutations work again

### LINKED ISSUE #74994: Mutations provided by enchantment silently fail to provide all of their traits
### Describe the bug

Testing something that involved one mutation granting other mutations through enchantment, and the result was that while the mutations were granted and showed up in the @ panel, most of their traits were not.

### Attach save file

[Haswell-trimmed.tar.gz](https://github.com/user-attachments/files/16196953/Haswell-trimmed.tar.gz)

### Steps to reproduce

1) Put the following JSON somewhere in your game so it loads.
```
{
    "type": "mutation",
    "id": "THE_TESTING_MUTATION",
    "name": { "str": "The Container Mutation" },
    "points": 99,
    "description": "You have extra mutations available to you.  Theoretically.",
    "valid": false,
    "starting_trait": false,
    "purifiable": false,
    "enchantments": [
      {
        "condition": "ALWAYS",
        "mutations": [ "FANGS", "BIRD_EYE" ]
      }
]}
```

2) Check the @ panel and see that you have the Fangs and Avian Eyes mutations
3) Check your inventory. You do not have the integrated fangs item that the Fangs mutation should provide.
4) Check your stats. Your perception is 8 and does not gain the +4 that Avian Eyes should provide. 

### Expected behavior

Mutations that show up in your @ panel provide their full benefits (and drawbacks). 

### Screenshots

image
image
image

The tilesets read the mutations at least:
image

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.22621.3737 (22H2)
- Game Version: cdda-experimental-2024-07-11-0546 e418d44 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth],
    No Rail Stations [No_Rail_Stations]
]

### Additional context

_No response_

### LINKED ISSUE #77971: Mutations' effects gained through passive effects of artifacts don't activate
### Describe the bug

Mutations' effects gained through passive effects of artifacts don't actually activate. While mutations are listed as in TRAITS tab of @-menu, they don't apply their effect to character.

### Attach save file

N/A

### Steps to reproduce

Debug-spawn clairvoyance artifact. Check @-menu. Notice that you have `Debug Clairvoyance Super` mutation, but there is no effect of that mutation. Drop the artifact and manually add the mutation. Notice that mutation effect is triggered immediately.

Debug spawn `Ixythyug the Destroyer's soul` from Magiclysm. Wear it. Check @-menu. Notice that you have lots of demonic mutations, but there is no effect of them, e.g. you don't have increased strength and dexterity from `Demonic Possession` mutation. If you manually add the mutation (without worn artifact), you'll get the increased stats.
Also it's worth mentioning that though you don't have increased strength from `Huge` (`DEMON_HUGE`) mutation gained by the same artifact, you still become huge and destroy your gear, this means that only some of effects are triggered, and some are not.

### Expected behavior

All mutations' effects gained through passive effects of artifacts should activate.

### Screenshots

изображение

изображение
Note that I have several demonic stat-changing mutations gained by the artifact, and no actual change in my stats.

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.5131 (22H2)
- Game Version: 0.G-14121-gfc8e17709b [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth],
    Magiclysm [magiclysm]
]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by osuphobia**
#### Summary
Infrastructure "Make enchantment mutations work again"
#### Purpose of change
Fix #74994 
Fix #77971 
#### Describe the solution
Removed dead codes about enchantment mutations.
Made new infrastructure, that save mutations to `cached_mutations`, allow functioning mutations to be properly read in `enchantment_cache`.
New enchantment mutations will override old enchantments and conflicting `my_mutations`, following the mutating rules. Overridden mutations are disabled untill said enchantments are removed.
Removed redundants caused by #74994 .
#### Testing
Worked fine and passed test cases locally.
Picking up The Architect's Cub and you can immediately see through walls.
Wear the Ixythyug the Destroyer's soul, and see that the bonus stats actually have effect, and the conflicting no longer works.
Transform to a bear, you got the traits, buffs, flags, integrated armors etc correctly, and can't transform into a cougar unless first back to human.
Change to cougar and see that you can pounce.
The mutation UI and @ menu also work as expected, character creating and debug mutating works as usual.

#### Additional context
Wearing the Ixythyug the Destroyer's soul immediately makes you emaciated, while druid transforming does not have this side effect. Is this intended behavior?

### CODE REVIEW

**Comment by Standing-Storm** [comment]
> Wearing the Ixythyug the Destroyer's soul immediately makes you emaciated, while druid transforming does not have this side effect. Is this intended behavior?

This is because your calories remain constant but your new size expects appropriate caloric values. In the druid forms, I have effects to directly modify your calories to account for the size changes (and return you to where they were before when you change back), whereas the Blood Mage spell just applies the mutations and has no such special handling. Becoming emaciated is a side effect of growing bigger so suddenly.

I did include flags to handle shapeshifted size, but actually making it account for calories was (and probably still is) beyond my skill level.

**Comment by SariusSkelrets** [comment]
It have special handling, but only if you gain the destroyer's soul through its spell. Spawing the item directly bypasses the handling EOC.

I know, I'm the one who added the demonic possession handling.

File: src/newcharacter.cpp [hunk 1]
```diff
@@ -4802,30 +4802,36 @@ std::vector<trait_id> Character::get_mutations( bool include_hidden,
...
-                for( const trait_and_var &exist : result ) {
-                    if( exist.trait == ench_trait ) {
-                        found = true;
-                        break;
-                    }
-                }
-                if( !found ) {
-                    result.emplace_back( ench_trait, "" );
-                }
+    auto &test = ignore_enchantments ? my_mutations : cached_mutations;
```
Comment (osuphobia) [inline src/newcharacter.cpp:4828]:
```suggestion
    const auto &test = ignore_enchantments ? my_mutations : cached_mutations;
```

File: src/newcharacter.cpp [hunk 2]
```diff
@@ -4802,30 +4802,36 @@ std::vector<trait_id> Character::get_mutations( bool include_hidden,
     return result;
 }
 
+std::vector<trait_id> Character::get_functioning_mutations( bool include_hidden,
+        bool ignore_enchantments, const std::function<bool( const mutation_branch & )> &filter ) const
+{
+    std::vector<trait_id> result;
+    auto &test = ignore_enchantments ? my_mutations : cached_mutations;
```
Comment (osuphobia) [inline src/newcharacter.cpp:4809]:
```suggestion
    const auto &test = ignore_enchantments ? my_mutations : cached_mutations;
```

File: src/magic_enchantment.h [hunk 3]
```diff
@@ -303,6 +303,8 @@ class enchant_cache : public enchantment
         void force_add( const enchant_cache &rhs );
         void force_add_with_dialogue( const enchantment &rhs, const const_dialogue &d,
                                       bool evaluate = true );
+        // adds enchantment mutations to the cache
+        void force_add_mutation( const enchantment &rhs, const Character &guy );
```
Comment (osuphobia) [inline src/magic_enchantment.h:307]:
```suggestion
        void force_add_mutation( const enchantment &rhs );
```

File: src/magic_enchantment.cpp [hunk 4]
```diff
@@ -777,6 +777,14 @@ bool enchant_cache::add( const enchant_cache &rhs )
     return true;
 }
 
+void enchant_cache::force_add_mutation( const enchantment &rhs, const Character &guy )
```
Comment (osuphobia) [inline src/magic_enchantment.cpp:780]:
```suggestion
void enchant_cache::force_add_mutation( const enchantment &rhs )
```

File: src/character.cpp [hunk 5]
```diff
@@ -7738,12 +7717,147 @@ void Character::recalculate_enchantment_cache()
         }
     }
 
+    for( const std::pair<const trait_id, trait_data> &mut_map : my_mutations ) {
+        if( mut_map.second.corrupted == 0 ) {
+            const mutation_branch &mut = mut_map.first.obj();
+            for( const enchantment_id &ench_id : mut.enchantments ) {
+                const enchantment &ench = ench_id.obj();
+                if( ench.is_active( *this, mut.activated && mut_map.second.powered ) ) {
+                    enchantment_cache->force_add_mutation( ench, *this );
```
Comment (osuphobia) [inline src/character.cpp:7726]:
```suggestion
                    enchantment_cache->force_add_mutation( ench );
```

**Comment by osuphobia** [comment]
> I did include flags to handle shapeshifted size (`TEMPORARY_SHAPESHIFT` etc) so that you are actually as big as you're supposed to be, but actually making a shapeshift size change automatically account for calories was (and probably still is) beyond my skill level.

I've seen this, and maybe we can use this flag to recalculate your body fat percentage:
```diff
 void Character::recalculate_size()
 {
     if( has_flag( json_flag_TEMPORARY_SHAPESHIFT ) ) {
-        if( has_flag( json_flag_SHAPESHIFT_SIZE_TINY ) ) {
-            size_class = creature_size::tiny;
-        } else if( has_flag( json_flag_SHAPESHIFT_SIZE_SMALL ) ) {
-            size_class = creature_size::small;
-        } else if( has_flag( json_flag_SHAPESHIFT_SIZE_LARGE ) ) {
-            size_class = creature_size::large;
-        } else if( has_flag( json_flag_SHAPESHIFT_SIZE_HUGE ) ) {
-            size_class = creature_size::huge;
-        } else {
-            size_class = creature_size::medium;
-        }
+        magical_resize();
     } else {
         if( has_flag( json_flag_TINY ) ) {
             size_class = creature_size::tiny;
@@ -545,6 +535,28 @@ void Character::recalculate_size()
     path_settings->size = size_class;
 }

+void Character::magical_resize()
+{
+    const float bmi_before = get_bmi_fat();
+    creature_size old_size_class = size_class;
+    if( has_flag( json_flag_SHAPESHIFT_SIZE_TINY ) ) {
+        size_class = creature_size::tiny;
+    } else if( has_flag( json_flag_SHAPESHIFT_SIZE_SMALL ) ) {
+        size_class = creature_size::small;
+    } else if( has_flag( json_flag_SHAPESHIFT_SIZE_LARGE ) ) {
+        size_class = creature_size::large;
+    } else if( has_flag( json_flag_SHAPESHIFT_SIZE_HUGE ) ) {
+        size_class = creature_size::huge;
+    } else {
+        size_class = creature_size::medium;
+    }
+    if( old_size_class != size_class ) {
+        const float bmi_after = get_bmi_fat();
+        int nkcal = get_stored_kcal() * bmi_before / bmi_after;
+        set_stored_kcal( nkcal );
+    }
+}
```

**Pull request merged**